File: dsa/topics/11_heaps/python/l295_find_median_from_data_stream.py
import heapq

class MedianFinder:
    """
    Problem: https://leetcode.com/problems/find-median-from-data-stream/

    Idea:
    Use 2 heaps.
    Store half elements in each heap.
    If odd number of elements, store one extra element in max heap.

    max_heap -> stores smaller half (as max heap)
    min_heap -> stores larger half (as min heap)

    Median can be derived from the top of heaps.
    """

    def __init__(self):
        self.max_heap = []   # max heap (store negatives)
        self.min_heap = []   # min heap

    def addNum(self, num: int) -> None:
        # Step 1: Always push to max heap first
        heapq.heappush(self.max_heap, -num)

        # Step 2: Balance order (max_heap top <= min_heap top)
        if self.min_heap and -self.max_heap[0] > self.min_heap[0]:
            heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))

        # Step 3: Balance size (max_heap can have at most 1 extra element)
        if len(self.max_heap) > len(self.min_heap) + 1:
            heapq.heappush(self.min_heap, -heapq.heappop(self.max_heap))
        elif len(self.min_heap) > len(self.max_heap):
            heapq.heappush(self.max_heap, -heapq.heappop(self.min_heap))

    def findMedian(self) -> float:
        if len(self.max_heap) == len(self.min_heap):
            return (-self.max_heap[0] + self.min_heap[0]) / 2.0
        return -self.max_heap[0]


if __name__ == "__main__":
    mf = MedianFinder()
    mf.addNum(1)
    mf.addNum(2)
    print(mf.findMedian())   # 1.5
    mf.addNum(3)
    print(mf.findMedian())   # 2

File: dsa/topics/11_heaps/java/L295_FindMedianFromDataStream.java
import java.util.Collections;
import java.util.PriorityQueue;
import java.util.Queue;

public class L295_FindMedianFromDataStream {

	public static void main(String[] args) {
		L295_FindMedianFromDataStream obj = new L295_FindMedianFromDataStream();

		obj.addNum(1);
		obj.addNum(2);
		System.out.println(obj.findMedian());
		obj.addNum(3);
		System.out.println(obj.findMedian());
	}

    // Problem: https://leetcode.com/problems/find-median-from-data-stream/
    // Idea: Use 2 pq's. Store half elements in each q. If odd element, store one extra in q1.
    // q1 should be max queue and q2 should be min queue.
    // You can find the median from top of the q elements.
	Queue<Integer> maxq;
    Queue<Integer> minq;
    public L295_FindMedianFromDataStream() {
        maxq = new PriorityQueue<>(Collections.reverseOrder());     // q1
        minq = new PriorityQueue<>();                               // q2
    }

    public void addNum(int num) {
        maxq.add(num);
        if(maxq.size()-minq.size() > 1)         // 1,2 -> this moves 2 to q2
            minq.add(maxq.remove());

        if(maxq.size() > 0 && minq.size() > 0){
            if(maxq.peek() > minq.peek()){      // 1,2,3 -> q1 will have 1,3 and q2 will have 2
                minq.add(maxq.remove());        // This will sync the q1 and q2
                maxq.add(minq.remove());        // 3 -> q2 and 2 -> q1
            }
        }
    }

    public double findMedian() {
        if(maxq.size() == minq.size()){
            return (maxq.peek()+minq.peek())/2.0;
        }else{
            return maxq.peek();
        }
    }
}
